### PR TITLE
Symbol table transformer

### LINKED ISSUE #291: Create visitor/transformer pattern for RMC/Goto structs
Going to be used for name transformation for --gen-c flag fix.

### PR DESCRIPTION
**Pull request by vecchiot-aws**
### Description of changes: 

Introduces a visitor pattern for transforming the symbol table. An example identity transformation is included to show the setup, and any of the default methods can be overridden to change behavior. Adds a compiler flag to list out symbol table passes to perform after codegenning.

### Resolved issues:

Resolves #291 

### Call-outs:

Todo:
- [x] Documentation

### Testing:

* How is this change tested? Existing regression suite, plus new unit test suite specific to the symbol table transformer trait.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:703]: Can the first line give a little more context on how this is used?

+ Nit: "explicitly"

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:1230]: Can we add a constructor that enforces this invariant?

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:16]: Nit: replace "it" with "the transformer" or similar specific item.

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:204]: Nit: spelling.

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:297]: Nit: single `=`?

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:712]: Add a comment here documenting when this is needed (the default symbol table stuff).

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:718]: Comment why this separate loop for None is needed.

**Review by avanhatt — CHANGES_REQUESTED** [review]
High level thoughts:

- Visitor pattern implementation seems clean, straightforward, and useful.
- Before landing, I would want to see some fairly heavy testing of this. Can we add a separate script that runs this over existing tests and checks that the identity is indeed correct? And that we can do simple changes, like inverting every boolean or similar?

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:711]: These two error messages are equal but the checks are different. Use different messages here.

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:714]: Same here.

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:100]: Shouldn't this be `transform_parameter_type` to match the other type-transforming function names?

**Review by adpaco-aws — COMMENTED** [review]
What about having the thing you are transforming be the second part of the function names?

For example, `transform_type_...` when transforming types, `transform_expr_...` when transforming expressions, and so on. Otherwise I feel like it's a bit harder to read/search, but maybe you've already considered this and decided this was the best way.

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:1230]: Maybe I should make the comment more clear, but this means manually wrapped in `.array_to_ptr()` *after* construction (the other constructor does it for you).

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:100]: It felt odd to give it that name since the parameter itself isn't a type (it's a data contained within some types), but for organization purposes that would makes sense.

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:703]: The dangers of copy-pasting existing comments...

Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:16]: Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:204]: Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:297]: Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:712]: I'm not sure what you mean. What's the default symbol table stuff?

Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:712]: Ohhh, the stuff already in there by default. Gotcha, fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:718]: Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:711]: Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:714]: Fixed!

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (vecchiot-aws) [reply compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs]: Fixed!

**Comment by vecchiot-aws** [comment]
>     * Before landing, I would want to see some fairly heavy testing of this. Can we add a separate script that runs this over existing tests and checks that the identity is indeed correct? And that we can do simple changes, like inverting every boolean or similar?

This has been manually checked, but I'm not sure what approach is best for automating it.

**Comment by avanhatt** [comment]
LGTM once some automated test is included in the PR (or, let me know if you feel like that should be a 2nd PR for timing).

**Review by avanhatt — APPROVED** [review]
LGTM!

**Review by adpaco-aws — APPROVED** [review]

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs
Comment (danielsn) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/transformer.rs:297]: Constructor doesn't exist?

**Review by danielsn — APPROVED** [review]
Might be nice to add something about this to a readme.

**Pull request merged**
